File: src/Core/Ast.fs
namespace Meticulous

/// The Abstract Syntax Tree (AST) for meticulous documents.
///
/// An AST is just a set of data types that describe the *shape* of a document
/// after we've read it. Parsing turns flat text into these structured values;
/// everything downstream (truth tables, rendering) works on the AST, never on
/// the raw text.
module Ast =

    /// A logical formula built from atomic propositions and connectives.
    /// This is the heart of the propositional layer. A `Formula` is a little
    /// tree — e.g. `p -> (q and r)` becomes `Implies(Atom "p", And(Atom "q", Atom "r"))`.
    type Formula =
        | Atom of string                    // a proposition variable, e.g. p
        | Const of bool                     // the constants ⊤ (true) and ⊥ (false)
        | Not of Formula                    // ¬φ
        | And of Formula * Formula          // φ ∧ ψ
        | Or of Formula * Formula           // φ ∨ ψ
        | Xor of Formula * Formula          // φ ⊕ ψ  (exclusive or)
        | Implies of Formula * Formula      // φ → ψ
        | Iff of Formula * Formula          // φ ↔ ψ  (if and only if)
        | Box of Formula                    // □φ — necessarily φ (true in every possible world)
        | Diamond of Formula                // ◇φ — possibly φ (true in some possible world)

    /// Where a `table` request points: either a named claim, or a formula
    /// written inline on the same line.
    type TableTarget =
        | TargetRef of string        // e.g.  table C1
        | TargetFormula of Formula   // e.g.  table (p -> q)

    /// The two kinds of `check` request.
    type CheckKind =
        | Verdict of Formula                 // check (p -> q)      -> is it a tautology?
        | Equivalent of Formula * Formula    // check A equivalent B -> are they the same?

    /// The structural relations a document can assert between statements.
    /// Three of them make formal claims the engine can verify; two are
    /// informal judgments it can only record.
    type RelationKind =
        | Supports      // informal: one statement lends weight to another
        | Presupposes   // informal: one statement silently relies on another
        | Contradicts   // formal: they can never both be true
        | Entails       // formal: whenever the first holds, so does the second
        | EquivalentTo  // formal: always the same truth value

    /// One end of a relation: a declared prop/claim name, or a quoted ad-hoc
    /// statement that hasn't been formalized ("The streets flood").
    type RelRef =
        | Named of string
        | Quoted of string

    /// One line of a user-written proof. Lines are numbered so later steps
    /// can cite the earlier ones they build on.
    type ProofLine =
        | ProofPremise of number: int * formula: Formula
                          // 1. premise rain -> wet
        | ProofDerived of number: int * formula: Formula * rule: string * refs: int list
                          // 3. wet   by modus-ponens from 1, 2

    /// A top-level statement — one meaningful line or block in a document.
    type Statement =
        | Heading of level: int * text: string   // # Title / ## Subtitle
        | Prose of string                          // any plain paragraph text
        | Prop of name: string * gloss: string     // prop p : It is raining
        | Claim of name: string * body: Formula    // claim C1 : p -> q
        | Table of TableTarget                      // table C1
        | Check of CheckKind                        // check ...
        | Argument of name: string * premises: Formula list * conclusion: Formula
                                                    // argument x { premise … / --- / conclude … }
        | Proof of name: string * lines: ProofLine list
                                                    // proof x { 1. premise … / 2. … by rule from 1 }
        | Analyze                                   // analyze — compare all claims pairwise
        | Relates of left: RelRef * kind: RelationKind * right: RelRef
                                                    // C1 supports C2 / A entails "the streets flood"
        | RelationMap                               // map — draw all asserted relations as a graph

    /// A whole document is simply an ordered list of statements.
    type Document = Statement list

File: vscode/src/core/Ast.js

import { Union } from "./fable_modules/fable-library-js.5.6.0/Types.js";
import { list_type, int32_type, union_type, bool_type, string_type } from "./fable_modules/fable-library-js.5.6.0/Reflection.js";

/**
 * A logical formula built from atomic propositions and connectives.
 * This is the heart of the propositional layer. A `Formula` is a little
 * tree — e.g. `p -> (q and r)` becomes `Implies(Atom "p", And(Atom "q", Atom "r"))`.
 */
export class Formula extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["Atom", "Const", "Not", "And", "Or", "Xor", "Implies", "Iff", "Box", "Diamond"];
    }
}

export function Formula_$reflection() {
    return union_type("Meticulous.Ast.Formula", [], Formula, () => [[["Item", string_type]], [["Item", bool_type]], [["Item", Formula_$reflection()]], [["Item1", Formula_$reflection()], ["Item2", Formula_$reflection()]], [["Item1", Formula_$reflection()], ["Item2", Formula_$reflection()]], [["Item1", Formula_$reflection()], ["Item2", Formula_$reflection()]], [["Item1", Formula_$reflection()], ["Item2", Formula_$reflection()]], [["Item1", Formula_$reflection()], ["Item2", Formula_$reflection()]], [["Item", Formula_$reflection()]], [["Item", Formula_$reflection()]]]);
}

/**
 * Where a `table` request points: either a named claim, or a formula
 * written inline on the same line.
 */
export class TableTarget extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["TargetRef", "TargetFormula"];
    }
}

export function TableTarget_$reflection() {
    return union_type("Meticulous.Ast.TableTarget", [], TableTarget, () => [[["Item", string_type]], [["Item", Formula_$reflection()]]]);
}

/**
 * The two kinds of `check` request.
 */
export class CheckKind extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["Verdict", "Equivalent"];
    }
}

export function CheckKind_$reflection() {
    return union_type("Meticulous.Ast.CheckKind", [], CheckKind, () => [[["Item", Formula_$reflection()]], [["Item1", Formula_$reflection()], ["Item2", Formula_$reflection()]]]);
}

/**
 * The structural relations a document can assert between statements.
 * Three of them make formal claims the engine can verify; two are
 * informal judgments it can only record.
 */
export class RelationKind extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["Supports", "Presupposes", "Contradicts", "Entails", "EquivalentTo"];
    }
}

export function RelationKind_$reflection() {
    return union_type("Meticulous.Ast.RelationKind", [], RelationKind, () => [[], [], [], [], []]);
}

/**
 * One end of a relation: a declared prop/claim name, or a quoted ad-hoc
 * statement that hasn't been formalized ("The streets flood").
 */
export class RelRef extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["Named", "Quoted"];
    }
}

export function RelRef_$reflection() {
    return union_type("Meticulous.Ast.RelRef", [], RelRef, () => [[["Item", string_type]], [["Item", string_type]]]);
}

/**
 * One line of a user-written proof. Lines are numbered so later steps
 * can cite the earlier ones they build on.
 */
export class ProofLine extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["ProofPremise", "ProofDerived"];
    }
}

export function ProofLine_$reflection() {
    return union_type("Meticulous.Ast.ProofLine", [], ProofLine, () => [[["number", int32_type], ["formula", Formula_$reflection()]], [["number", int32_type], ["formula", Formula_$reflection()], ["rule", string_type], ["refs", list_type(int32_type)]]]);
}

/**
 * A top-level statement — one meaningful line or block in a document.
 */
export class Statement extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["Heading", "Prose", "Prop", "Claim", "Table", "Check", "Argument", "Proof", "Analyze", "Relates", "RelationMap"];
    }
}

export function Statement_$reflection() {
    return union_type("Meticulous.Ast.Statement", [], Statement, () => [[["level", int32_type], ["text", string_type]], [["Item", string_type]], [["name", string_type], ["gloss", string_type]], [["name", string_type], ["body", Formula_$reflection()]], [["Item", TableTarget_$reflection()]], [["Item", CheckKind_$reflection()]], [["name", string_type], ["premises", list_type(Formula_$reflection())], ["conclusion", Formula_$reflection()]], [["name", string_type], ["lines", list_type(ProofLine_$reflection())]], [], [["left", RelRef_$reflection()], ["kind", RelationKind_$reflection()], ["right", RelRef_$reflection()]], []]);
}

